stm32-modules/heater-shaker/simulator/sim_driver.cpp

```diff
@@ -0,0 +1,101 @@
+#include "simulator/sim_driver.hpp"
+
+#include <boost/asio.hpp>
+#include <regex>
+
+#include "simulator/simulator_queue.hpp"
+
+using namespace sim_driver;
+
+sim_driver::SocketSimDriver::SocketSimDriver(std::string url) {
+    std::regex url_regex(":\\/\\/([a-zA-Z0-9.]*):(\\d*)$");
+    std::smatch url_match_result;
+
+    if (std::regex_search(url, url_match_result, url_regex)) {
+        host = url_match_result[1];
+        port = std::stoi(url_match_result[2]);
+
+    } else {
+        std::cerr << "Malformed url." << std::endl;
+        exit(1);
+    }
+}
+
+std::string sim_driver::SocketSimDriver::get_host() { return this->host; }
+
+int sim_driver::SocketSimDriver::get_port() { return this->port; }
+
+std::string sim_driver::SocketSimDriver::get_name() { return this->name; }
+
+void sim_driver::SocketSimDriver::write() {}
+
+auto get_socket(std::string host, int port) {
```

Comment:
Consider the relative lifetimes of the socket and the `SocketSimDriver` object. It's reasonable to tie the acquisition of the socket for `bind()` to the lifetime of the `SocketSimDriver`; this is a c++ pattern known as [RAII](https://en.cppreference.com/w/cpp/language/raii), "resource acquisition is initialization". 

It becomes a little more complex for receive-side sockets (more on that later) because a receive-side BSD socket has _two_ lifetimes associated with it:
- the `bind()` lifetime, which lasts while the program can accept incoming connections (or, rather, is ready to handle incoming connection _requests_ - it's allowed to reject them while maintaining the `bind()`)
- the `listen()` lifetime, which is the same as the lifetime of an individual connection.

The typical pattern is to have a socket manager own the bind, and create and spin off new objects for each connection.

Here, we'll only have one connection - we're emulating a single physical wire, after all - so I don't think we need two objects, but the different lifetimes still have different semantics, especially since there could be a connection that starts, and ends, and then restarts, etc. 

A good pattern might be to essentially replace `get_socket` with a class that RAII's a bind-socket, so you have like
```
class Socket {
   explicit Socket(AddrInfo addr) {
      sock = bind(addr.host, addr.port);
      if(!sock) { // or whatever error signalling mechanism
        throw std::runtime_error("Could not acquire socket ");
      }
   }
   ~Socket() {
        sock.destroy(); // or whatever teardown
    }
   Socket sock;
};
```

At this point, you can replace a call to `get_socket` with just creating a `Socket`, and you can trust that whenever the function ends, whether through return or exception, this object will get destroyed (because it's created by-value in the function) and the socket will get cleaned up.

You can then also have methods to listen or whatever.
and replace calls to `get_socket` with just building s.a clas